diff --git a/Makefile.pre.in b/Makefile.pre.in
index a7dc970..63c9367 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -471,6 +471,8 @@ PYTHON_OBJS=	\
 		Python/pyhash.o \
 		Python/pylifecycle.o \
 		Python/pymath.o \
+		Python/pystandalone.o \
+		Python/pystandalone_payload.o \
 		Python/pystate.o \
 		Python/pythonrun.o \
 		Python/pytime.o \
@@ -499,6 +501,9 @@ PYTHON_OBJS=	\
 		$(DTRACE_OBJS) \
 		@PLATFORM_OBJS@
 
+Python/pystandalone_payload.o: Python/pystandalone_payload.S
+	$(CC) $(PY_CORE_CFLAGS) -c $< -o $@
+
 
 ##########################################################################
 # Objects
@@ -2488,7 +2493,7 @@ TESTSUBDIRS=	idlelib/idle_test \
 		test/xmltestdata/c14n-20 \
 		test/zipimport_data
 
-COMPILEALL_OPTS=-j0
+COMPILEALL_OPTS=-j1
 
 TEST_MODULES=@TEST_MODULES@
 
