diff --git i/Makefile.pre.in w/Makefile.pre.in
index 7ca3dc6..1ceb4a8 100644
--- i/Makefile.pre.in
+++ w/Makefile.pre.in
@@ -411,6 +411,8 @@ PYTHON_OBJS=	\
 		Python/pyhash.o \
 		Python/pylifecycle.o \
 		Python/pymath.o \
+		Python/pystandalone.o \
+		Python/pystandalone_payload.o \
 		Python/pystate.o \
 		Python/pythonrun.o \
 		Python/pytime.o \
@@ -437,6 +439,9 @@ PYTHON_OBJS=	\
 		$(DTRACE_OBJS) \
 		@PLATFORM_OBJS@

+Python/pystandalone_payload.o: Python/pystandalone_payload.S
+	$(CC) $(PY_CORE_CFLAGS) -c $< -o $@
+

 ##########################################################################
 # Objects
