diff --git i/Makefile.pre.in w/Makefile.pre.in
index 81d4d50..c084f58 100644
--- i/Makefile.pre.in
+++ w/Makefile.pre.in
@@ -111,7 +111,7 @@ NO_AS_NEEDED=	@NO_AS_NEEDED@
 CCSHARED=	@CCSHARED@
 # LINKFORSHARED are the flags passed to the $(CC) command that links
 # the python executable -- this is only needed for a few systems
-LINKFORSHARED=	@LINKFORSHARED@
+LINKFORSHARED=	@LINKFORSHARED@ -sectcreate __PYSTANDALONE __pystandalone Python/pystandalone_payload
 ARFLAGS=	@ARFLAGS@
 # Extra C flags added for building the interpreter object files.
 CFLAGSFORSHARED=@CFLAGSFORSHARED@
@@ -414,6 +414,8 @@ PYTHON_OBJS=	\
 		Python/pyhash.o \
 		Python/pylifecycle.o \
 		Python/pymath.o \
+		Python/pystandalone.o \
+		Python/pystandalone_payload \
 		Python/pystate.o \
 		Python/pythonrun.o \
 		Python/pytime.o \
@@ -438,6 +440,9 @@ PYTHON_OBJS=	\
 		$(DTRACE_OBJS) \
 		@PLATFORM_OBJS@
 
+Python/pystandalone_payload:
+	dd if=/dev/zero of=$@ bs=1m count=4
+
 
 ##########################################################################
 # Objects
